File: BOJ/16XXX/16234-인구_이동.js
/**
 * ABOUT
 *
 * DATE: 2021-08-10
 * AUTHOR: lovelysi0113
 *
 * URL: https://www.acmicpc.net/problem/16234
 *
 * COMMENT:
 * 인구 이동을 계속 진행하며 각 칸의 인구 수를 갱신해준다
 **/

const readline = require('readline');

const rl = readline.createInterface({
	input: process.stdin,
	output: process.stdout,
});

let input = [];

rl.on('line', function (line) {
	input.push(line.split(' ').map(el => parseInt(el)));
}).on('close', function () {
	let [N, L, R] = input[0];
	let map = input.slice(1);

	const DELTAS = [
		[0, 1],
		[0, -1],
		[1, 0],
		[-1, 0],
	];

	// bfs를 사용하여 인구 이동 진행
	const bfs = () => {
		// 방문한 칸 표시
		let visited = Array.from(Array(N), () => Array(N).fill(false));
		// 인구 이동이 진행되었는지 확인
		let isMoved = false;

		// 인구 이동 시작
		for (let x = 0; x < N; x++) {
			for (let y = 0; y < N; y++) {
				if (!visited[x][y]) {
					// 방문할 칸 저장
					let queue = [];
					queue.push([x, y]);
					visited[x][y] = true;
					// 국경선을 연 칸들의 목록 - 연합
					let open = [[x, y]];
					let total = map[x][y];
					while (queue.length > 0) {
						let [nx, ny] = queue.shift();
						// 인접한 칸과 인구 차이를 비교하여 L~R 사이면 국경선 열기
						DELTAS.forEach(([dx, dy]) => {
							let [chkx, chky] = [nx + dx, ny + dy];
							if (0 <= chkx && chkx < N && 0 <= chky && chky < N && !visited[chkx][chky]) {
								let differ = Math.abs(map[chkx][chky] - map[nx][ny]);
								if (L <= differ && differ <= R) {
									// 국경선 열기
									queue.push([chkx, chky]);
									open.push([chkx, chky]);
									visited[chkx][chky] = true;
									isMoved = true;
									total += map[chkx][chky];
								}
							}
						});
					}
					// 연합을 이루고 있는 각 칸의 인구수 변경
					open.forEach(([ox, oy]) => {
						map[ox][oy] = Math.floor(total / open.length);
					});
				}
			}
		}
		return isMoved;
	};

	// 인구 이동 계속 반복하기
	let moveCount = 0;
	while (true) {
		if (bfs()) {
			moveCount += 1;
		} else {
			break;
		}
	}

	console.log(moveCount);

	process.exit();
});

File: BOJ/16XXX/16234-인구_이동.py
'''/**
 * ABOUT
 *
 * DATE: 2021-08-11
 * AUTHOR: lovelysi0113
 *
 * URL: https://www.acmicpc.net/problem/16234
 *
 * COMMENT:
 * 인구 이동을 계속 진행하며 각 칸의 인구 수를 갱신해준다
 **/'''

from collections import deque

DELTAS = [(0, 1), (0, -1), (1, 0), (-1, 0)]

N, L, R = map(int, input().split(' '))
ground = [list(map(int, input().split(' '))) for _ in range(N)]

DELTAS = [(0, 1), (0, -1), (1, 0), (-1, 0)]
# bfs를 사용하여 인구 이동 진행
def bfs():
    # 방문한 칸 표시
    visited = [[False] * N for _ in range(N)]
    # 인구 이동이 진행되었는지 확인
    is_moved = False;

    # 인구 이동 시작
    for x in range(N):
        for y in range(N):
            if not visited[x][y]:
                # 방문할 칸 저장
                queue = deque()
                queue.append((x, y))
                visited[x][y] = True
                # 국경선을 연 칸들의 목록 - 연합
                open_group = [(x, y)]
                total = ground[x][y]
                while queue:
                    nx, ny = queue.popleft()
                    # 인접한 칸과 인구 차이를 비교하여 L~R 사이면 국경선 열기
                    for dx, dy in DELTAS:
                        chkx, chky = nx + dx, ny + dy
                        if 0 <= chkx < N and 0 <= chky < N and not visited[chkx][chky]:
                            differ = abs(ground[chkx][chky] - ground[nx][ny])
                            if L <= differ <= R:
                                # 국경선 열기
                                queue.append((chkx, chky))
                                open_group.append((chkx, chky))
                                visited[chkx][chky] = True
                                is_moved = True
                                total += ground[chkx][chky]
                # 연합을 이루고 있는 각 칸의 인구수 변경
                value = total // len(open_group)
                for ox, oy in open_group:
                    ground[ox][oy] = value
    return is_moved

# 인구 이동 계속 반복하기
move_count = 0
while True:
    if bfs():
        move_count += 1
    else:
        break

print(move_count)